### PR TITLE
Addresser Registry

### PR DESCRIPTION
**Pull request by adamgering**
Add address parse, deserialization, get_address_type, and
an addresser registry to facilitate root level addresser
methods for these functions.

Also adds some container name hints.

Task for Issue #210 and #552

### CODE REVIEW

File: rbac/common/base/base_address.py [hunk 1]
```diff
@@ -98,15 +192,22 @@ def relationship_type(self):
         if it is an address type that stores relationships"""
         raise NotImplementedError("Class must implement this property")
 
+    @property
+    def address_type_name(self):
+        """Returns a unique name for this addresser based on the combination
+        of the properties object_type + related_type + relationship_type"""
```
Comment (nadiabahrami) [inline rbac/common/base/base_address.py]: Were we adding email into the address?

File: rbac/common/role/relationship_admin.py [hunk 2]
```diff
@@ -56,3 +56,13 @@ def related_type(self):
...
         """The related type from AddressSpace implemented by this class"""
         return addresser.RelationshipType.ADMIN
+
+    @property
+    def _state_object_name(self):
+        """Task relationship state object name ends with Relationship (TaskRelationship)"""
+        return self._name_camel + "Relationship"
+
+    @property
+    def _state_container_list_name(self):
```
Comment (nadiabahrami) [inline rbac/common/role/relationship_admin.py]: Why do you have the same function returning a string "relationship" in two different classes?  Here

File: rbac/common/role/relationship_member.py [hunk 3]
```diff
@@ -56,3 +56,13 @@ def related_type(self):
...
         """The related type from AddressSpace implemented by this class"""
         return addresser.RelationshipType.MEMBER
+
+    @property
+    def _state_object_name(self):
+        """Task relationship state object name ends with Relationship (TaskRelationship)"""
+        return self._name_camel + "Relationship"
+
+    @property
+    def _state_container_list_name(self):
```
Comment (nadiabahrami) [inline rbac/common/role/relationship_member.py]: And Here

File: rbac/common/role/relationship_owner.py [hunk 4]
```diff
@@ -56,3 +56,13 @@ def related_type(self):
...
         return addresser.RelationshipType.OWNER
+
+    @property
+    def _state_object_name(self):
+        """Task relationship state object name ends with Relationship (TaskRelationship)"""
+        return self._name_camel + "Relationship"
+
+    @property
+    def _state_container_list_name(self):
+        """Tasks relationship state container collection name is relationships"""
```
Comment (nadiabahrami) [inline rbac/common/role/relationship_owner.py]: and here

File: rbac/common/role/relationship_task.py [hunk 5]
```diff
@@ -56,3 +56,13 @@ def related_type(self):
...
         """The related type from AddressSpace implemented by this class"""
         return addresser.RelationshipType.MEMBER
+
+    @property
+    def _state_object_name(self):
+        """Task relationship state object name ends with Relationship (TaskRelationship)"""
+        return self._name_camel + "Relationship"
+
+    @property
+    def _state_container_list_name(self):
```
Comment (nadiabahrami) [inline rbac/common/role/relationship_task.py]: and here.   I feel like this is a good use for base classing.  Why are you not adding _state_object_name and _state_container_list_name to the base class BaseRelationship?  Then there would not be as many repeats

File: rbac/common/base/base_address.py (hunk 1 above)
Comment (adamgering) [reply rbac/common/base/base_address.py]: Addresses are not related to emails

File: rbac/common/base/base_address.py [hunk 6]
```diff
@@ -116,5 +217,38 @@ def address_is(self, address):
...
+        return None
+
+    def get_object_id(self, address):
+        """Returns the hash of an object_id (or the object_id itself if it
+        is a 12-byte unique identifier), as encoded in a given address"""
+        return address[14:38]
+
+    def get_target_id(self, address):
+        """Returns the hash of a target_id (or the target_id itself if it
+        is a 12-byte unique identifier), as encoded in a given address"""
```
Comment (kthblmfld) [inline rbac/common/base/base_address.py]: This comment doesn't line up with the behavior of the function.

File: rbac/common/proposal/proposal_address.py [hunk 7]
```diff
@@ -40,6 +44,11 @@ def relationship_type(self):
         """The related type from AddressSpace implemented by this class"""
         return addresser.RelationshipType.ATTRIBUTES
 
+    @property
+    def _state_container_prefix(self):
+        """Proposal container is plural (ProposalsContainer)"""
+        return self._state_object_name + "s"
```
Comment (kthblmfld) [inline rbac/common/proposal/proposal_address.py]: Should this happen in init()? It seems odd to be concatenating an 's' onto another property on its way out. Maybe an additional property?

Fundamentally, the idea is that data should flow through the system without alteration except in obvious places (ie transformers). If you are going to keep this, then the docstring should at least make it clear what it is doing internally.

File: rbac/common/base/base_address.py (hunk 6 above)
Comment (adamgering) [reply rbac/common/base/base_address.py]: What specifically?

File: rbac/common/proposal/proposal_address.py (hunk 7 above)
Comment (adamgering) [reply rbac/common/proposal/proposal_address.py]: It’s a property that can be overridden.
It’s a hint for the class to know the state data schema implemented by that address type.

File: rbac/common/base/base_address.py (hunk 6 above)
Comment (adamgering) [reply rbac/common/base/base_address.py]: """Returns the hash of a target_id (or the target_id itself if it
        is a 12-byte unique identifier), as encoded in a given address"""

Yes, get_target_id does correctly return the hash of the target_id (or the target_id itself it is a 12-byte unique identifier), as encoded in the address.

    @pytest.mark.parse_role_member_address
    def test_parse_role_member_address_hash(self):
        """Test addresser.user.parse returns a parsed address if it is a user address"""
        user_id = "foo"
        role_id = "bar"
        address = addresser.role.member.address(role_id, user_id)
        parsed = addresser.role.member.parse(address)

        self.assertEqual(parsed.object_type, addresser.ObjectType.ROLE)
        self.assertEqual(parsed.related_type, addresser.ObjectType.USER)
        self.assertEqual(
            parsed.relationship_type, addresser.RelationshipType.MEMBER
        )
        self.assertEqual(parsed.address_type, addresser.AddressSpace.ROLES_MEMBERS)
        self.assertEqual(parsed.object_id, addresser.role.member.hash(role_id))
        self.assertEqual(parsed.target_id, addresser.role.member.hash(user_id))

**Pull request merged**